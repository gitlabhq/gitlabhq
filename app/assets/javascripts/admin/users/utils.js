import { queryToObject } from '~/lib/utils/url_utility';
import getSoloOwnedOrganizationsQuery from '~/admin/users/graphql/queries/get_solo_owned_organizations.query.graphql';
import { convertToGraphQLId } from '~/graphql_shared/utils';
import { TYPENAME_USER } from '~/graphql_shared/constants';
import {
  TOKEN_CONFIGS,
  SOLO_OWNED_ORGANIZATIONS_EMPTY,
  SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT,
} from 'ee_else_ce/admin/users/constants';
import { OPERATOR_IS } from '~/vue_shared/components/filtered_search_bar/constants';

export const generateUserPaths = (paths, id) => {
  return Object.fromEntries(
    Object.entries(paths).map(([action, genericPath]) => {
      return [action, genericPath.replace('id', id)];
    }),
  );
};

/**
 * @typedef {{type: string, value: {data: string, operator: string}}} Token
 */

/**
 * Initialize token values based on the URL parameters
 *
 * @returns {{tokenValues: Array<string|Token>, sort: string}}
 */
export function initializeValuesFromQuery() {
  const tokenValues = [];

  const { filter, search_query: searchQuery, sort } = queryToObject(window.location.search);

  if (filter) {
    const tokenConfig = TOKEN_CONFIGS.find(({ options }) =>
      options.some(({ value }) => value === filter),
    );

    if (tokenConfig) {
      tokenValues.push({
        type: tokenConfig.type,
        value: { data: filter, operator: OPERATOR_IS },
      });
    }
  }

  if (searchQuery) {
    tokenValues.push(searchQuery);
  }

  return { tokenValues, sort };
}

export const getSoloOwnedOrganizations = async (apolloClient, userId) => {
  if (!window.gon?.features?.uiForOrganizations) {
    return Promise.resolve(SOLO_OWNED_ORGANIZATIONS_EMPTY);
  }

  const {
    data: {
      user: { organizations },
    },
  } = await apolloClient.query({
    query: getSoloOwnedOrganizationsQuery,
    variables: {
      id: convertToGraphQLId(TYPENAME_USER, userId),
      first: SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT,
    },
  });

  return organizations;
};
