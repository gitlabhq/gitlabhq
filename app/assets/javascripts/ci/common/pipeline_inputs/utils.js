import { isEqual } from 'lodash';

/**
 * Generates skeleton rect props for the skeleton loader based on column and row indices
 *
 * @param {Number} columnIndex - The column index (0-based)
 * @param {Number} rowIndex - The row index (0-based)
 * @returns {Object} - The props for the skeleton rect
 */
export const getSkeletonRectProps = (columnIndex, rowIndex) => {
  return {
    x: `${columnIndex * 25.5}%`,
    y: rowIndex * 10,
    width: '23%',
    height: 6,
    rx: 2,
    ry: 2,
  };
};

/**
 * Formats a value for display in the input preview
 * @param {*} value - The value to format
 * @returns {string} - Formatted string representation
 */
export function formatValue(value) {
  if (value === null) return 'null';
  if (value === undefined) return 'undefined';
  if (typeof value === 'string') return `"${value}"`;
  if (typeof value === 'object') return JSON.stringify(value);
  return String(value);
}

/**
 * Checks if an input value has changed from its default
 * @param {*} value - Current value
 * @param {*} defaultValue - Default value
 * @returns {boolean} - True if values are different
 */
export function hasValueChanged(value, defaultValue) {
  if (typeof value === 'object' || typeof defaultValue === 'object') {
    return !isEqual(value, defaultValue);
  }
  return value !== defaultValue;
}

/**
 * Formats value lines for diff display
 * @param {Object} input - Input object with value, default, etc.
 * @param {boolean} isChanged - Whether the value has changed
 * @returns {Array} - Array of line objects
 */
export function formatValueLines(input, isChanged) {
  const lines = [];
  const formattedValue = formatValue(input.value);
  const formattedDefault = formatValue(input.default);

  if (isChanged) {
    lines.push({
      // eslint-disable-next-line @gitlab/require-i18n-strings
      content: `-   value: ${formattedDefault}`,
      type: 'old',
    });
    lines.push({
      // eslint-disable-next-line @gitlab/require-i18n-strings
      content: `+   value: ${formattedValue}`,
      type: 'new',
    });
  } else {
    lines.push({
      // eslint-disable-next-line @gitlab/require-i18n-strings
      content: `    value: ${formattedValue}`,
      type: '',
    });
  }

  return lines;
}

/**
 * Formats metadata lines (type, description)
 * @param {Object} input - Input object
 * @returns {Array} - Array of line objects
 */
export function formatMetadataLines(input) {
  const lines = [];

  if (input.type) {
    lines.push({
      // eslint-disable-next-line @gitlab/require-i18n-strings
      content: `    type: "${input.type}"`,
      type: '',
    });
  }

  if (input.description) {
    lines.push({
      // eslint-disable-next-line @gitlab/require-i18n-strings
      content: `    description: "${input.description}"`,
      type: '',
    });
  }

  return lines;
}

/**
 * Formats a single input into display lines
 * @param {Object} input - Input object
 * @returns {Array} - Array of line objects
 */
export function formatInputLines(input) {
  const lines = [];
  const isChanged = hasValueChanged(input.value, input.default);

  lines.push({
    content: `${input.name}:`,
    type: '',
  });

  lines.push(...formatValueLines(input, isChanged));

  lines.push(...formatMetadataLines(input));

  lines.push({
    content: '',
    type: '',
  });

  return lines;
}

/**
 * Formats all inputs into display lines
 * @param {Array} inputs - Array of input objects
 * @returns {Array} - Array of line objects for display
 */
export function formatInputsForDisplay(inputs) {
  return inputs.flatMap(formatInputLines);
}
