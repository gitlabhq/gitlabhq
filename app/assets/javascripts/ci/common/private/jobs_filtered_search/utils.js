import { JOB_SOURCES } from 'ee_else_ce/ci/common/private/jobs_filtered_search/tokens/constants';
import { jobStatusValues, jobRunnerTypeValues } from './constants';

// validates query string used for filtered search
// on jobs table to ensure GraphQL query is called correctly
export const validateQueryString = (queryStringObj) => {
  return Object.entries(queryStringObj).reduce((acc, [queryStringKey, queryStringValue]) => {
    switch (queryStringKey) {
      case 'statuses': {
        const statusValue = queryStringValue.toUpperCase();
        const statusValueValid = jobStatusValues.includes(statusValue);
        return statusValueValid ? { ...acc, statuses: statusValue } : acc;
      }
      case 'sources': {
        const sourceValue = queryStringValue.toUpperCase();
        const sourceValueValid = JOB_SOURCES.some((source) => source.value === sourceValue);
        return sourceValueValid ? { ...acc, sources: sourceValue } : acc;
      }
      case 'runnerTypes': {
        const runnerTypesValue = queryStringValue.toUpperCase();
        const runnerTypesValueValid = jobRunnerTypeValues.includes(runnerTypesValue);
        return runnerTypesValueValid ? { ...acc, runnerTypes: runnerTypesValue } : acc;
      }
      case 'kind': {
        const jobKindValue = queryStringValue.toUpperCase();
        const jobKindValueValid = ['BUILD', 'BRIDGE'].includes(jobKindValue);
        return jobKindValueValid ? { ...acc, kind: jobKindValue } : acc;
      }
      case 'name': {
        return { ...acc, name: queryStringValue };
      }
      default:
        return acc;
    }
  }, null);
};
