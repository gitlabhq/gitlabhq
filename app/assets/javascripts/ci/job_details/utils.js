export const compactJobLog = (jobLog) => {
  const compactedLog = [];

  jobLog.forEach((obj) => {
    // push header section line
    if (obj.line && obj.isHeader) {
      compactedLog.push(obj.line);
    }

    // push lines within section header
    if (obj.lines?.length > 0) {
      compactedLog.push(...obj.lines);
    }

    // push lines from plain log
    if (!obj.lines && obj.content.length > 0) {
      compactedLog.push(obj);
    }
  });

  return compactedLog;
};

export const filterAnnotations = (annotations, type) => {
  return [...annotations]
    .sort((a, b) => a.name.localeCompare(b.name))
    .flatMap((annotationList) => annotationList.data)
    .flatMap((annotation) => annotation[type] ?? []);
};

export const getLineText = (line) => {
  return line.content.map((content) => content.text).join('');
};
