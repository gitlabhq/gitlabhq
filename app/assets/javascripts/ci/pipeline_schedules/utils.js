export const updateScheduleNodes = (schedules = [], updatedSchedule = {}) => {
  return schedules.map((schedule) => {
    if (schedule.id === updatedSchedule.id) {
      return {
        ...schedule,
        lastPipeline: updatedSchedule.lastPipeline,
      };
    }

    return schedule;
  });
};
