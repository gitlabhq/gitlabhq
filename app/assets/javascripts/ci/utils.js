import * as Sentry from '~/sentry/sentry_browser_wrapper';

export const reportToSentry = (component, error) => {
  Sentry.captureException(error, {
    tags: {
      component,
    },
  });
};
