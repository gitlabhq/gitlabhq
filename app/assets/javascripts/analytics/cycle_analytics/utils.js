import { parseSeconds, newDate } from '~/lib/utils/datetime_utility';
import { formatTimeAsSummary } from '~/lib/utils/datetime/date_format_utility';
import { joinPaths } from '~/lib/utils/url_utility';

/**
 * Takes the stages and median data, combined with the selected stage, to build an
 * array which is formatted to proivde the data required for the path navigation.
 *
 * @param {Array} stages - The stages available to the group / project
 * @param {Object} medians - The median values for the stages available to the group / project
 * @param {Object} stageCounts - The total item count for the stages available
 * @param {Object} selectedStage - The currently selected stage
 * @returns {Array} An array of stages formatted with data required for the path navigation
 */
export const transformStagesForPathNavigation = ({
  stages,
  medians,
  stageCounts = {},
  selectedStage,
}) => {
  const formattedStages = stages.map((stage) => {
    return {
      metric: medians[stage?.id],
      selected: stage?.id && stage?.id === selectedStage?.id,
      stageCount: stageCounts && stageCounts[stage?.id],
      icon: null,
      ...stage,
    };
  });

  return formattedStages;
};

/**
 * Takes a raw median value in seconds and converts it to a string representation
 * ie. converts 172800 => 2d (2 days)
 *
 * @param {Number} Median - The number of seconds for the median calculation
 * @returns {String} String representation ie 2w
 */
export const medianTimeToParsedSeconds = (value) =>
  formatTimeAsSummary({
    ...parseSeconds(value, { daysPerWeek: 7, hoursPerDay: 24 }),
    seconds: value,
  });

/**
 * Takes the raw median value arrays and converts them into a useful object
 * containing the string for display in the path navigation
 * ie. converts [{ id: 'test', value: 172800 }] => { 'test': '2d' }
 *
 * @param {Array} Medians - Array of stage median objects, each contains a `id`, `value` and `error`
 * @returns {Object} Returns key value pair with the stage name and its display median value
 */
export const formatMedianValues = (medians = []) =>
  medians.reduce((acc, { id, value = 0 }) => {
    return {
      ...acc,
      [id]: value ? medianTimeToParsedSeconds(value) : '-',
    };
  }, {});

export const filterStagesByHiddenStatus = (stages = [], isHidden = true) =>
  stages.filter(({ hidden = false }) => hidden === isHidden);

/**
 * Builds the initial data object for Value Stream Analytics with data loaded from the backend
 *
 * @param {Object} dataset - dataset object paseed to the frontend via data-* properties
 * @returns {Object} - The initial data to load the app with
 */
export const buildCycleAnalyticsInitialData = ({
  projectId,
  groupPath,
  stage,
  createdAfter,
  createdBefore,
  namespaceName,
  namespacePath,
  namespaceRestApiRequestPath,
} = {}) => {
  return {
    projectId: parseInt(projectId, 10),
    groupPath,
    namespace: {
      name: namespaceName,
      path: namespacePath,
      restApiRequestPath: namespaceRestApiRequestPath,
    },
    createdAfter: newDate(createdAfter),
    createdBefore: newDate(createdBefore),
    selectedStage: stage ? JSON.parse(stage) : null,
  };
};

export const constructPathWithNamespace = ({ restApiRequestPath }, endpoint) =>
  joinPaths('/', restApiRequestPath, endpoint);
