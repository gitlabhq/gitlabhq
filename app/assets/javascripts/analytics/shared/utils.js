import dateFormat from '~/lib/dateformat';
import { SECONDS_IN_DAY } from '~/lib/utils/datetime_utility';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';
import { joinPaths } from '~/lib/utils/url_utility';
import { urlQueryToFilter } from '~/vue_shared/components/filtered_search_bar/filtered_search_utils';
import {
  dateFormats,
  FLOW_METRICS,
  MAX_METRIC_PRECISION,
  UNITS,
  VALUE_STREAM_METRIC_DISPLAY_UNITS,
  VALUE_STREAM_METRIC_METADATA,
  VALUE_STREAM_METRIC_TILE_METADATA,
} from './constants';

export const filterBySearchTerm = (data = [], searchTerm = '', filterByKey = 'name') => {
  if (!searchTerm?.length) return data;
  return data.filter((item) => item[filterByKey].toLowerCase().includes(searchTerm.toLowerCase()));
};

export const toYmd = (date) => dateFormat(date, dateFormats.isoDate);

/**
 * Takes a url and extracts query parameters used for the shared
 * filter bar
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractFilterQueryParameters = (url = '') => {
  const {
    source_branch_name: selectedSourceBranch = null,
    target_branch_name: selectedTargetBranch = null,
    author_username: selectedAuthor = null,
    milestone_title: selectedMilestone = null,
    assignee_username: selectedAssigneeList = [],
    label_name: selectedLabelList = [],
  } = urlQueryToFilter(url);

  return {
    selectedSourceBranch,
    selectedTargetBranch,
    selectedAuthor,
    selectedMilestone,
    selectedAssigneeList,
    selectedLabelList,
  };
};

/**
 * Takes a url and extracts sorting and pagination query parameters into an object
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractPaginationQueryParameters = (url = '') => {
  const { sort, direction, page } = urlQueryToFilter(url);
  return {
    sort: sort?.value || null,
    direction: direction?.value || null,
    page: page?.value || null,
  };
};

export const getDataZoomOption = ({
  totalItems = 0,
  maxItemsPerPage = 40,
  dataZoom = [{ type: 'slider', bottom: 10, start: 0 }],
}) => {
  if (totalItems <= maxItemsPerPage) {
    return {};
  }

  const intervalEnd = Math.ceil((maxItemsPerPage / totalItems) * 100);

  return dataZoom.map((item) => {
    return {
      ...item,
      end: intervalEnd,
    };
  });
};

/**
 * Formats any valid number as percentage
 *
 * @param {number|string} decimalValue Decimal value between 0 and 1 to be converted to a percentage
 * @param {number} precision The number of decimal places to round to
 *
 * @returns {string} Returns a formatted string multiplied by 100
 */
export const formatAsPercentageWithoutSymbol = (decimalValue = 0, precision = 1) => {
  const parsed = Number.isNaN(Number(decimalValue)) ? 0 : decimalValue;
  return (parsed * 100).toFixed(precision);
};

/**
 * Converts a time in seconds to number of days, with variable precision
 *
 * @param {Number} seconds Time in seconds
 * @param {Number} precision Specifies the number of digits after the decimal
 *
 * @returns {String} The number of days
 */
export const secondsToDays = (seconds, precision = 1) =>
  (seconds / SECONDS_IN_DAY).toFixed(precision);

export const scaledValueForDisplay = (value, units, precision = MAX_METRIC_PRECISION) => {
  switch (units) {
    case UNITS.PERCENT:
      return formatAsPercentageWithoutSymbol(value);
    case UNITS.DAYS:
      return secondsToDays(value, precision);
    default:
      return value;
  }
};

const prepareMetricValue = ({ identifier, value, unit }) => {
  // NOTE: the flow metrics graphql endpoint returns values already scaled for display
  if (!value) {
    // ensures we return `-` for 0/null etc
    return '-';
  }
  return Object.values(FLOW_METRICS).includes(identifier)
    ? value
    : scaledValueForDisplay(value, unit);
};

/**
 * Prepares metric data to be rendered in the metric_tile component
 *
 * @param {MetricData[]} data - The metric data to be rendered
 * @returns {TransformedMetricData[]} An array of metrics ready to render in the metric_tile
 */
export const rawMetricToMetricTile = (metric) => {
  const { identifier, value, ...metricRest } = metric;
  const { unit, label, ...metadataRest } = VALUE_STREAM_METRIC_TILE_METADATA[identifier];
  return {
    ...metadataRest,
    ...metricRest,
    title: label,
    identifier,
    label,
    unit: VALUE_STREAM_METRIC_DISPLAY_UNITS[unit],
    value: prepareMetricValue({ value, unit, identifier }),
  };
};

/**
 * Generates a URL link to the VSD dashboard based on the
 * namespace path passed into the method.
 *
 * @param {String} namespacePath - Path of the specified namespace
 * @param {Boolean} isProjectNamespace
 * @returns a URL or blank string if there is no namespacePath set
 */
export const generateValueStreamsDashboardLink = (
  namespacePath = null,
  isProjectNamespace = false,
) => {
  if (!namespacePath) return '';

  const dashboardsSlug = '/-/analytics/dashboards/value_streams_dashboard';
  const formattedNamespacePath = isProjectNamespace ? namespacePath : `groups/${namespacePath}`;
  const segments = [gon.relative_url_root || '', '/', formattedNamespacePath, dashboardsSlug];

  return joinPaths(...segments);
};

/**
 * Generates a URL link to the metric's report based on the namespace type.
 * @param {String} metricId â€“ Metric's identifier
 * @param {String} namespacePath - Path of the specified namespace
 * @param {Boolean} isProjectNamespace
 * @returns {String} URL link to the metric's report
 */
export const generateMetricLink = ({
  metricId = null,
  namespacePath = null,
  isProjectNamespace = false,
} = {}) => {
  const { groupLink, projectLink } = VALUE_STREAM_METRIC_METADATA[metricId] ?? {};
  const hasLinks = groupLink || projectLink;

  if (!hasLinks || !namespacePath) return '';

  return joinPaths(
    '/',
    gon.relative_url_root,
    !isProjectNamespace ? 'groups' : '',
    namespacePath,
    isProjectNamespace ? projectLink : groupLink,
  );
};

/**
 * Extracts the relevant feature and license flags needed for VSA
 *
 * @param {Object} gon the global `window.gon` object populated when the page loads
 * @returns an object containing the extracted feature flags and their boolean status
 */
export const extractVSAFeaturesFromGON = () => ({
  // licensed feature toggles
  cycleAnalyticsForGroups: Boolean(gon?.licensed_features?.cycleAnalyticsForGroups),
  cycleAnalyticsForProjects: Boolean(gon?.licensed_features?.cycleAnalyticsForProjects),
  groupLevelAnalyticsDashboard: Boolean(gon?.licensed_features?.groupLevelAnalyticsDashboard),
});

/**
 * Takes a raw GraphQL response which could contain data for a group or project namespace,
 * and returns the data for the namespace which is present in the response.
 *
 * @param {Object} params
 * @param {string} params.resultKey - The data to be extracted from the namespace.
 * @param {Object} params.result
 * @param {Object} params.result.data
 * @param {Object} params.result.data.group - The group GraphQL response.
 * @param {Object} params.result.data.project - The project GraphQL response.
 * @returns {Object} The data extracted from either group[resultKey] or project[resultKey].
 */
export const extractQueryResponseFromNamespace = ({ result, resultKey }) => {
  const { group = null, project = null } = result.data;
  if (group || project) {
    const namespace = group ?? project;
    return namespace[resultKey] || {};
  }
  return {};
};

/**
 * Takes the raw snake_case query parameters and extracts + converts the relevant values
 * for the overview metrics component
 * @param {Object} params - Object containing the supported query parameters
 * @param {Date} params.created_before
 * @param {Date} params.created_after
 * @param {string} params.author_username
 * @param {string} params.milestone_title
 * @param {Array} params.label_name
 * @param {Array} params.assignee_username
 *
 * @returns {Object} CamelCased parameter names
 */
export const overviewMetricsRequestParams = (params = {}) => {
  const {
    createdAfter: startDate,
    createdBefore: endDate,
    labelName: labelNames,
    assigneeUsername: assigneeUsernames,
    ...rest
  } = convertObjectPropsToCamelCase(params);
  return {
    startDate,
    endDate,
    labelNames,
    assigneeUsernames,
    ...rest,
  };
};
