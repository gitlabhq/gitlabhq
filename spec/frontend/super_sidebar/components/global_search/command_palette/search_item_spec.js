import { GlAvatar } from '@gitlab/ui';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import SearchItem from '~/super_sidebar/components/global_search/command_palette/search_item.vue';
import { getFormattedItem } from '~/super_sidebar/components/global_search/utils';
import { linksReducer } from '~/super_sidebar/components/global_search/command_palette/utils';
import { USERS, LINKS, PROJECT, ISSUE } from './mock_data';

jest.mock('~/lib/utils/highlight', () => ({
  __esModule: true,
  default: (text) => text,
}));
const mockUser = getFormattedItem(USERS[0]);
const mockCommand = LINKS.reduce(linksReducer, [])[1];
const mockProject = getFormattedItem(PROJECT);
const mockIssue = getFormattedItem(ISSUE);

describe('SearchItem', () => {
  let wrapper;

  const createComponent = (item) => {
    wrapper = shallowMountExtended(SearchItem, {
      propsData: {
        item,
        searchQuery: 'root',
      },
    });
  };

  const findAvatar = () => wrapper.findComponent(GlAvatar);
  const findIcon = () => wrapper.findByTestId('icon');
  const findNamespace = () => wrapper.findByTestId('namespace');
  const findNamespaceBullet = () => wrapper.findByTestId('namespace-bullet');

  it.each([mockUser, mockCommand, mockProject, mockIssue])('should render the item', (item) => {
    createComponent(item);

    expect(wrapper.element).toMatchSnapshot();
  });

  describe('item rendering', () => {
    it('should render avatar when avatar_url is provided', () => {
      const item = {
        text: 'Test Item',
        avatar_url: 'https://example.com/avatar.png',
        entity_id: '123',
        entity_name: 'Test Entity',
      };

      createComponent(item);

      expect(findAvatar().props('src')).toBe(item.avatar_url);
      expect(findAvatar().props('entityId')).toBe(item.entity_id);
      expect(findAvatar().props('entityName')).toBe(item.entity_name);
      expect(findAvatar().props('size')).toBe(16);
      expect(findAvatar().attributes('aria-hidden')).toBe('true');
    });

    it('should not render avatar when avatar_url is undefined', () => {
      const item = {
        text: 'Test Item',
      };

      createComponent(item);

      expect(findAvatar().exists()).toBe(false);
    });

    it('should render icon when present', () => {
      const item = {
        icon: 'search-results',
        text: 'Test Item',
      };

      createComponent(item);

      expect(findIcon().props('name')).toBe(item.icon);
    });

    it('should not render icon when not present', () => {
      const item = {
        text: 'Test Item',
      };

      createComponent(item);

      expect(findIcon().exists()).toBe(false);
    });

    it('should render namespace when present', () => {
      const item = {
        text: 'Test Item',
        namespace: 'test-namespace',
      };

      createComponent(item);

      expect(findNamespaceBullet().exists()).toBe(true);
      expect(findNamespace().text()).toBe('test-namespace');
    });

    it('should not render namespace when not present', () => {
      const item = {
        text: 'Test Item',
      };

      createComponent(item);

      expect(findNamespaceBullet().exists()).toBe(false);
      expect(findNamespace().exists()).toBe(false);
    });
  });
});
