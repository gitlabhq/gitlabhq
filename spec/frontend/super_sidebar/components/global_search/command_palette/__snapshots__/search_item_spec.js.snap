// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchItem should render the item 1`] = `
<div
  class="gl-flex gl-items-center"
>
  <gl-avatar-stub
    alt="avatar"
    aria-hidden="true"
    class="gl-mr-3"
    entityid="37"
    entityname=""
    shape="rect"
    size="16"
    src="https://www.gravatar.com/avatar/a9638f4ec70148d51e56bf05ad41e993?s=80&d=identicon"
  />
  <span
    class="gl-flex gl-gap-2 gl-items-center gl-min-w-0"
  >
    <span
      class="gl-text-strong gl-truncate"
    >
      Cole Dickinson
    </span>
  </span>
</div>
`;

exports[`SearchItem should render the item 2`] = `
<div
  class="gl-flex gl-items-center"
>
  <gl-icon-stub
    class="gl-mr-3 gl-shrink-0"
    data-testid="icon"
    name="users"
    size="16"
    variant="current"
  />
  <span
    class="gl-flex gl-gap-2 gl-items-center gl-min-w-0"
  >
    <span
      class="gl-text-strong gl-truncate"
    >
      Manage &gt; Activity
    </span>
  </span>
</div>
`;

exports[`SearchItem should render the item 3`] = `
<div
  class="gl-flex gl-items-center"
>
  <gl-avatar-stub
    alt="avatar"
    aria-hidden="true"
    class="gl-mr-3"
    entityid="1"
    entityname="MockProject1"
    shape="rect"
    size="16"
    src="/project/avatar/1/avatar.png"
  />
  <span
    class="gl-flex gl-gap-2 gl-items-center gl-min-w-0"
  >
    <span
      class="gl-text-strong gl-truncate"
    >
      MockProject1
    </span>
    <span
      aria-hidden="true"
      class="gl-text-subtle"
      data-testid="namespace-bullet"
    >
      Â·
    </span>
    <span
      class="gl-text-sm gl-text-subtle gl-truncate"
      data-testid="namespace"
    >
      Gitlab Org / MockProject1
    </span>
  </span>
</div>
`;

exports[`SearchItem should render the item 4`] = `
<div
  class="gl-flex gl-items-center"
>
  <gl-avatar-stub
    alt="avatar"
    aria-hidden="true"
    class="gl-mr-3"
    entityid="7"
    entityname="Flight"
    shape="rect"
    size="16"
    src=""
  />
  <span
    class="gl-flex gl-gap-2 gl-items-center gl-min-w-0"
  >
    <span
      class="gl-text-strong gl-truncate"
    >
      Dismiss Cipher with no integrity
    </span>
  </span>
</div>
`;
